### PR TITLE
mostekcm add client deploy capability

### PR DESCRIPTION
**Pull request by mostekcm**
This adds a fairly basic ability to add and update clients.  It won't do anything if you don't pass in any clients.

### CODE REVIEW

**Comment by Verlic** [comment]
MD5: c4103f122d27677c9db144cae1394a66

# NSP - Security Report

_This is an automated security audit of this project. Please do not modify its contents_

Vulnerabilities found: **2**
Affected Modules: **uglify-js, uglify-js**
Summary:

- Regular Expression Denial of Service
- Incorrect Handling of Non-Boolean Comparisons During Minification


## Detailed report
#### Regular Expression Denial of Service

Affected Module: **uglify-js**
Installed version: **2.2.5**
Patched versions: **>=2.6.0**
Advisory: https://nodesecurity.io/advisories/48
Path: **auth0-source-control-extension-tools@2.0.0 > auth0-oauth2-express@1.2.0 > jade@1.11.0 > transformers@2.1.0 > uglify-js@2.2.5**

---

#### Incorrect Handling of Non-Boolean Comparisons During Minification

Affected Module: **uglify-js**
Installed version: **2.2.5**
Patched versions: **>= 2.4.24**
Advisory: https://nodesecurity.io/advisories/39
Path: **auth0-source-control-extension-tools@2.0.0 > auth0-oauth2-express@1.2.0 > jade@1.11.0 > transformers@2.1.0 > uglify-js@2.2.5**

---

File: src/auth0/clients.js [hunk 1]
```diff
@@ -0,0 +1,267 @@
+const _ = require('lodash');
+const Promise = require('bluebird');
+const ValidationError = require('auth0-extension-tools').ValidationError;
+
+const utils = require('../utils');
+// const constants = require('../constants');
```
Comment (sandrinodimattia) [inline src/auth0/clients.js]: Remove the commented line

File: src/auth0/clients.js [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        .flattenDeep()
+        .union()
+        .value();
+      return progress.clients;
+    });
+};
+
+const updateExistingClient = function(progress, client, clientName, clientConfig, existingClient) {
+  /* Get entire client config */
+  clientConfig = utils.parseJsonFile(clientName, clientConfig.scriptFile);
```
Comment (sandrinodimattia) [inline src/auth0/clients.js]: does `scriptFile` make sense here?  can we rename to something else? like `settingsFile` `configurationFile` ... ?

File: src/auth0/clients.js [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+/**
+ * Make sure the management client exists and is filtered out, also make sure we have scriptFile defined for everything
+ * @param clients the list of clients to add or update
+ * @param managementClient the name of the management client
+ * @param existingClients the list of clients that already have been added
+ * @returns {Promise.<*>}
+ */
+const validateClientsExistence = function(clients, managementClient, existingClients) {
+  let existingClientsFiltered = existingClients;
```
Comment (sandrinodimattia) [inline src/auth0/clients.js]: Using `let` will cause issues in webtask. can you change this to `var`?

File: src/deploy.js [hunk 4]
```diff
@@ -27,14 +30,15 @@ const trackProgress = function(progressData) {
   };
 };
 
-module.exports = function(progressData, context, client, storage, config, slackTemplate) {
+module.exports = function(progressData, context, client, storage, config, slackTemplate, managementClientId) {
```
Comment (sandrinodimattia) [inline src/deploy.js]: Do we need to specify this? We should be able to retrieve it using `config('AUTH0_CLIENT_ID')` no?

File: src/auth0/clients.js (hunk 2 above)
Comment (mostekcm) [inline src/auth0/clients.js]: Yes, I was being lazy so I could re-use one of the utils.  I'll clean it up.

File: src/auth0/clients.js (hunk 1 above)
Comment (mostekcm) [inline src/auth0/clients.js]: done

File: src/deploy.js (hunk 4 above)
Comment (mostekcm) [inline src/deploy.js]: Good point, I started with passing the name, then changed it and didn't notice config was passed in.

File: src/auth0/clients.js (hunk 1 above)
Comment (mostekcm) [inline src/auth0/clients.js]: done

File: src/auth0/clients.js (hunk 2 above)
Comment (mostekcm) [inline src/auth0/clients.js]: Good point, I've fixed this too.

File: src/auth0/clients.js (hunk 3 above)
Comment (mostekcm) [inline src/auth0/clients.js]: done

File: src/deploy.js (hunk 4 above)
Comment (mostekcm) [inline src/deploy.js]: good catch, I was passing in name originally, then changed to ID and missed that config was passed in!

Comment (mostekcm) [inline src/deploy.js]: done

**Pull request merged**